File: 2021_22 - Practicum in Object Oriented Programming/Olympic Games Statistics App/GUI App/Java/src/olympics/competitors/Athlete.java
package olympics.competitors;

import java.util.Objects;

import olympics.events.Event;


public class Athlete extends Competitor {
	
	// ==========================================
	// Attributes:


	private Integer id;
	private String name;
	private char gender = '-';  // M/F = male/female, dash for not defined
	private int age = -1;		 // -1 for not defined
	private int height = -1;    // -1 for not defined
	private float weight = -1.f;    // -1 for not defined


	// ==========================================
    // Constructors:


	public Athlete(final int id) {
		this.id = id;
	}

	public Athlete(final int id, Event event_,
				   final int oly_games_id)
	{
		this(id);
		this.addEvent(event_, oly_games_id);
	}

    
	// ==========================================
	// Getters:
	
	
	public Integer getId() {
		return id;
	}

	public String getName() {
		return name;
	}

	public char getGender() {
		return gender;
	}

	public int getAge() {
		return age;
	}

	public int getHeight() {
		return height;
	}

	public float getWeight() {
		return weight;
	}

	
	// ==========================================
	// Setters:
	
	
	public void setId(Integer id) {
		this.id = id;
	}

	public void setName(String name) {
		this.name = name;
	}

	public void setGender(char gender) {
		this.gender = gender;
	}

	public void setAge(int age) {
		this.age = age;
	}

	public void setHeight(int height) {
		this.height = height;
	}

	public void setWeight(float weight) {
		this.weight = weight;
	}


	// ==========================================
	// Hashing:


	@Override
	public int hashCode() {
		return Objects.hash(id);
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (getClass() != obj.getClass())
			return false;
		Athlete other = (Athlete) obj;
		return id == other.id;
	}

}

File: 2021_22 - Practicum in Object Oriented Programming/Olympic Games Statistics App/CLI App/src/Athlete.cpp
#include "Athlete.hpp"


// ================================
// Constructors:


sale::Athlete::Athlete(const int id)
	: id(id)
{
}

sale::Athlete::Athlete(const int id,
					   std::shared_ptr<sale::Event> event_ptr,
					   const int oly_games_id)
	: id(id)
{
	this->addEvent(std::move(event_ptr), oly_games_id);
}


// ================================
// Public Setters:


void sale::Athlete::setName(const std::string& name)
{
	this->name = name;
}

void sale::Athlete::setGender(const char gender)
{
	this->gender = gender;
}

void sale::Athlete::setAge(const int age)
{
	this->age = age;
}

void sale::Athlete::setHeight(const int height)
{
	this->height = height;
}

void sale::Athlete::setWeight(const float weight)
{
	this->weight = weight;
}

const std::string& sale::Athlete::getNameReference() const
{
	return this->name;
}

std::string sale::Athlete::getName() const
{
	return this->name;
}

int sale::Athlete::getAge() const
{
	return this->age;
}

int sale::Athlete::getHeight() const
{
	return this->height;
}

float sale::Athlete::getWeight() const
{
	return this->weight;
}


// ================================
// HashableData overrides:


sale::Athlete::TKey sale::Athlete::getKey() const
{
	return this->id;
}


// ================================
// Operator Overloads:


bool operator==(const sale::Athlete& x, const sale::Athlete& y)
{
	return x.id == y.id;
}


namespace sale{


std::ostream& operator<<(std::ostream& os, const sale::Athlete& self)
{
	return os << self.getNameReference() << " (" << self.getKey() << ")";
}


}  // end of sale namespace
